### ISSUE TITLE
Deal with popup that exceeds maxBounds

### ISSUE THREAD

**Issue by martinheidegger** [Leaflet/Leaflet#2324]
If you open a popup that exceeds the bounds defined via the maxBounds option then the popup will be cut on top. There should be some mechanism that offers a easy way to deal with this problem. I can imagine:
- when testing for bounds limit include the popup size
- render the popup differently (on the bottom, left or right) if it exceeds the maxBounds.
- Add mode that lets popups break out of the map viewport.

**Comment by webyak** [comment]
+1 I think this is crucial when working with image overlays

**Comment by IvanSanchez** [comment]
I think this is covered by the Rrose plugin (http://erictheise.github.io/rrose/), so there would be no need to implement this in the core.

**Comment by mourner** [comment]
Well, it's not super-critical but we'd still want to fix that because it's not just a feature, but a bug you can trigger quite easily and that harms accessibility.

**Comment by jessertaylor** [comment]
Anyone able to look at this pretty please? :+1:

**Comment by mourner** [comment]
@jessertaylor just use the RRose plugin for now.

**Comment by jessertaylor** [comment]
@mourner I've looked it at but was unsure if I can get it working with https://github.com/tombatossals/angular-leaflet-directive or not - I will give it a try.

**Comment by Karpfador** [comment]
Not sure if my issue is similar enough but:
FlyTo() does not seem to respect maxBounds. 

When I animate focusin on one marker on my map (which is inside bounds), then flyTo will still move the map to have said marker centered. Then the map bounces back to fix itself according to max bounds. It would be great if the function could calculate an offset for the animation when maxbounds are set

**Comment by fdenzer** [comment]
So, two years later we still have to use Rrose - which is not as good a solution as having this fixes in Core module.
Leaflet is used as an ecosystem of third party components. It is annoying to monkey-patch Vue2Leaflet, or Angular components, with Rrose, just to give two examples.

**Comment by beneuto** [comment]
My suggestion is for options as follows:

1. Popup underneath if marker latitude is south of centre latitude
2. When autopanning, allow bounds to be exceeded if required to accommodate popup.   When popup closes go back to map centre as before opened .

**Comment by SebastianZ** [comment]
To stay consistent, I suggest to add a `direction` option like [there is for `Tooltip`](https://leafletjs.com/reference.html#tooltip-direction). The `auto` value then positions the popup so that it always stays within the `maxBounds`.

Or, if possible, move it to [`DivOverlay`](https://leafletjs.com/reference.html#divoverlay), so all overlays based on it can benefit from that functionality.

Sebastian
